sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -394,42 +384,53 @@ public String getEventHubName() {
         final String partitionKey = batch.getPartitionKey();
         final boolean isTracingEnabled = tracerProvider.isEnabled();
         final AtomicReference<Context> parentContext = isTracingEnabled
-            ? new AtomicReference<>(Context.NONE)
-            : null;
+                                                           ? new AtomicReference<>(Context.NONE)
+                                                           : null;
+        final AtomicReference<Context> sharedContext = isTracingEnabled
+                                                           ? new AtomicReference<>(Context.NONE)
+                                                           : null;
+        final AtomicReference<Boolean> isFirst = new AtomicReference<>(true);
 
         final List<Message> messages = batch.getEvents().stream().map(event -> {
```

Comment:
Consider using a `for` loop rather than the `stream().map()` so you don't need the "isFirst" variable and the parentContext and sharedContext do not have to be atomicreferences.

Reply:
I don't think we can avoid the `isFirst`  because we don't want to send the request to `getSharedSpanBuilder` for every event, we can just do that for one event and get the shared builder for all others. 

Also, `parentContext ` would still have to be an AtomicReference as it is used [here](https://github.com/Azure/azure-sdk-for-java/pull/6773/files#diff-afcf12d6b6264dbb96ae63baef89082dR423) in this lambda.

Reply:
I'm fairly confident this works:

```java
Context parentContext = null;
Context sharedContext = null;
List<Message> messages = new ArrayList<>();

for (int = 0; i < batch.getEvents().size(); i++) {
    final EventData = batch.getEvents().get(i);
    if (isTracingEnabled) {
        parentContext = event.getContext();
        if (i == 0) {
            sharedContext = tracerProvider.getSharedSpanBuilder(parentContext.get());
        }

        Context sharedSpanBuilderContext = sharedContext.get();
        tracerProvider.addSpanLinks(sharedSpanBuilderContext.addData(SPAN_CONTEXT_KEY, event.getContext()));
    }

    final Message message = messageSerializer.serialize(event);
    messages.add(message);
}
```

Reply:
We are using sharedContext and parentContext in the lambda expressions below and so they need to be `effectively final`.